src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+its subpatterns against the list's elements to destructure them by position.
+
+List patterns require that the pattern match the entire list, but you can use
+[rest elements](#rest-element) as a place holder.
+
+### Rest element
+
+List patterns can contain _one_ rest element (`...`) which allows matching lists
+of arbitrary lengths.
+
+```dart
+var [a, b, ..., c, d] = [1, 2, 3, 4, 5, 6, 7];
+print('$a $b $c $d'); // Prints "1 2 6 7".
+```
+
+Rest elements can also have subpatterns that collects elements that don't match
+the other subpatterns in the list, into a new list:
+
+```dart
+var [a, b, ...rest, c, d] = [1, 2, 3, 4, 5, 6, 7];
+print('$a $b $rest $c $d'); // Prints "1 2 [3, 4, 5] 6 7".
+```
+
+## Logical-and	
+
+`subpattern1 && subpattern2`
+
+A pair of patterns separated by `&&` matches only if both subpatterns match. If the
+left branch does not match, the right branch is not evaluated. 
+
+## Logical-or
+
+`subpattern1 || subpattern2`
+
+A pair of patterns separated by `||` matches if either of the branches match.
+If the left branch matches, the right branch is not evaluated. 
+
+Logical-or patterns can be nested inside a destructuring pattern. 
+
+```dart
+switch (list) {
+  // Matches a two-element list whose first element is 'a' or 'b':
+  case ['a' || 'b', var c]:
+}
+```
+
+{% comment %}
+[TODO: move the below content to the switch page, doesn't belong here]
+{% endcomment %}
+
+Logical-or patterns are useful for having multiple cases share a body in switch
+expressions or statements.
+
+```dart
+var isPrimary = switch (color) {
+  Color.red || Color.yellow || Color.blue => true,
+  _ => false
+};
+```
+
+Switch statements can have multiple cases share a body without using logical-or
+patterns, but they are still uniquely useful for allowing multiple cases to share
+a guard: 
+
+```dart
+switch (shape) {
+  case Square(size: var s) || Circle(size: var s) when s > 0:
+    print('Non-empty symmetric shape');
+}
+```
+
+## Map
+
+`{"key": subpattern1, someConst: subpattern2}`
+
+Map patterns match values that implement [`Map`][], and then recursively 
+match its subpatterns against the map’s keys to destructure them.
+
+Map patterns don't require the pattern to match the entire map. If a map has
+extra keys that te pattern doesn't destructure, it can still match.
+
+## Null-assert	
+
+`subpattern!`
+
+Null-assert patterns permit non-null values to flow through, but [throw][] if the
+matched value is null. 
+
+```dart
+List<String?> row = // ...
+
+// If the first column is 'user', we expect to have a name after it.
+switch (row) {
+  case ['user', var name!]:
+    // name is a non-nullable string here.
+}
+```
+
+## Null-check	
+
+`subpattern?`
+
+Null-check patterns match if the value is not null, and then match the inner
+pattern against that same value.
+
+
+```dart
+String? maybeString = // ...
+switch (maybeString) {
+  case var s?:
+    // s has type non-nullable String here.
+}
+```
```

Comment:
It might be worth pointing out that you use the constant pattern `null` to match when the value _is_ null. Someone asked me about this recently.